### PR TITLE
[Serve] Add multiplex support

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- Introduce multiplex API.
1. @serve.multiplexed(num_models_per_replica=0)
2. @serve.get_model_id

**router & controller & deployment_state change will be in followed up pr.**

Usage:
```
@serve.deployment
class Model:

    @serve.multiplexed(num_models_per_replica=1)
    def get_model(self, tag):
        print(f"load model {tag}")
        return tag

    async def __call__(self, request):
        model_id = serve.get_model_id()
        fake_model = await self.get_model(model_id)
        return fake_model()
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -305,6 +337,32 @@ def __init__(
             call_in_event_loop=event_loop,
         )
 
+        self.deployment_name = deployment_name
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:340]: No change, move the queue handle metrics pusher to Router.

File: python/ray/serve/_private/utils.py [hunk 2]
```diff
@@ -535,3 +536,27 @@ def record_serve_tag(key: str, value: str):
         )
 
     record_extra_usage_tag(serve_telemetry_tag_map[key], value)
+
+
+def _extract_self_if_method_call(args: List[Any], func: Callable) -> Optional[object]:
```
Comment (sihanwang41) [inline python/ray/serve/_private/utils.py:541]: No change move the function out of the batching.py

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -141,11 +148,9 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
         """Try to assign query to a replica, return the object ref if succeeded
         or return None if it can't assign this query to any replicas.
         """
-        for _ in range(len(self.in_flight_queries.keys())):
-            replica = next(self.replica_iterator)
-            if len(self.in_flight_queries[replica]) >= replica.max_concurrent_queries:
-                # This replica is overloaded, try next one
-                continue
+
+        def _assign_replica(replica: RunningReplicaInfo):
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:152]: Could you make this a separate, external method instead of a nested method?

File: python/ray/serve/_private/router.py [hunk 4]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
                 )
                 self.in_flight_queries[replica].add(tracker_ref)
             return user_ref
+
+        # Try to find a replica that can handle this query.
+        # If model id is not specified, we can assign it to any non-overladed replica.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:192]:
```suggestion
        # If model id is not specified, we can assign it to any non-overloaded replica.
```

File: python/ray/serve/_private/router.py [hunk 5]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
+        # If model id is specified, we can try to assign it to a replica that has
+        # loaded, and if there is no such replica, we can assign it to any
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:194]:
```suggestion
        # loaded the model, and if there is no such replica, we can assign it to any
```

File: python/ray/serve/_private/router.py [hunk 6]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
+        # non-overloaded replica.
+        if query.metadata.model_id:
+            # Try to find all replicas that has loaded the model and is not overloaded.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:197]:
```suggestion
            # Try to find a replica that has loaded the model and is not overloaded.
```

File: python/ray/serve/_private/router.py [hunk 7]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
+            candidate_replicas = [
+                replica
+                for replica in self.in_flight_queries.keys()
+                if query.metadata.model_id in replica.model_ids
+                and len(self.in_flight_queries[replica])
+                < replica.max_concurrent_queries
+            ]
+            for replica in self.in_flight_queries.keys():
+                logger.info(f"router debug: {replica.replica_tag}, {replica.model_ids}")
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:206]: Should this be a debug log?

File: python/ray/serve/_private/router.py [hunk 8]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
+            if candidate_replicas:
+                # If there are such replicas, randomly pick one.
+                replica = random.choice(candidate_replicas)
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:209]: Why do we do random choice instead of round robin across the candidate replicas?

File: python/ray/serve/_private/router.py [hunk 9]
```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
...
             return user_ref
+
+        # Try to find a replica that can handle this query.
+        # If model id is not specified, we can assign it to any non-overladed replica.
+        # If model id is specified, we can try to assign it to a replica that has
+        # loaded, and if there is no such replica, we can assign it to any
+        # non-overloaded replica.
+        if query.metadata.model_id:
+            # Try to find all replicas that has loaded the model and is not overloaded.
+            candidate_replicas = [
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:198]: This policy seems to work best if a new model is loaded due to only a single request.

However, suppose a few different end users requests a new model all at once. Ideally, one replica should load that model and serve the few users, just like what happens if they send their requests one after another. Instead, those requests will be sent to different HTTP proxies, and those proxies will assign the requests to many different replicas (since no replica currently has the model loaded). All those replicas may evict their current model when loading the new model. This could cause some thrashing if there’s a couple groups of end users, each requesting a different model.

In other words, if n clients request a new model at roughly the same time, n replicas will evict their current model and load the new one. Ideally, if n is small, only 1 replica should load the model and serve all n clients.

I think this is a relatively common usage pattern, so we should be aware of it. I'm not sure if it's worth handling right now versus in the future.

File: python/ray/serve/api.py [hunk 10]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
     """
     client = get_global_client()
     client.delete_apps([name], blocking=_blocking)
+
+
+@PublicAPI(stability="alpha")
+def multiplexed(func=None, num_models_per_replica: int = 0):
+    """Coverts a function or method to a multiplexed function.
+
+    The function can be standalone function or a method of a class. The
```
Comment (shrekris-anyscale) [inline python/ray/serve/api.py:586]: Can you type hint `func` with `Callable[[str], None]`? (I'm assuming the return type is `None`).

File: python/ray/serve/api.py [hunk 11]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+                model_multiplex_wrapper = getattr(multiplex_object, multiplex_attr)
+            return await model_multiplex_wrapper.load_model(*args, **kwargs)
+
+        return _multiplex_wrapper
+
+    return _multiplex_decorator(func) if callable(func) else _multiplex_decorator
+
+
+@PublicAPI(stability="alpha")
+def get_model_id() -> str:
```
Comment (shrekris-anyscale) [inline python/ray/serve/api.py:671]: Do we need to support this as a public API?

File: python/ray/serve/controller.py [hunk 12]
```diff
@@ -179,6 +179,14 @@ def record_autoscaling_metrics(self, data: Dict[str, float], send_timestamp: flo
     def record_handle_metrics(self, data: Dict[str, float], send_timestamp: float):
         self.deployment_state_manager.record_handle_metrics(data, send_timestamp)
 
+    def record_model_ids(self, data: Tuple[str, str, Tuple[str]]):
+        """Record model ids for a replica of deployment
```
Comment (shrekris-anyscale) [inline python/ray/serve/controller.py:183]:
```suggestion
        """Record model ids for a deployment replica.
```

File: python/ray/serve/multiplex.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+
+class _ModelMultiplexWrapper:
+    def __init__(self, model_load_func, self_args, num_models_per_replica=0):
+        # The models are stored in an OrderedDict to ensure LRU caching.
+        self.models = OrderedDict()
+        self._func = sync_to_async(model_load_func)
+        self.self_args = self_args
+        self.num_models_per_replica = num_models_per_replica
+        self.model_load_latency_s = metrics.Gauge(
```
Comment (shrekris-anyscale) [inline python/ray/serve/multiplex.py:28]: Could you add a metric to track number of evictions by `model_id`?

File: python/ray/serve/tests/test_multiplex.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
+import pytest
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_multiplex.py:1]: Could you add a test with an object that has an explicit `__del__` method to check that the method gets executed?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work so far! I left some comments.

File: python/ray/serve/_private/http_proxy.py [hunk 15]
```diff
@@ -421,12 +422,20 @@ async def __call__(self, scope, receive, send):
...
 
-        start_time = time.time()
+        model_id = ""
+        for key, value in scope["headers"]:
+            if RAY_SERVE_REQUEST_MODEL_ID == key.decode():
+                model_id = value.decode()
         ray.serve.context._serve_request_context.set(
             ray.serve.context.RequestContext(
-                route_path, get_random_letters(10), app_name
+                route_path,
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/http_proxy.py:431]: yes, that is the basic case we want to support one model_id per deployment. There is no customer requirements to support multiple multiplexed deployment within one request.

File: python/ray/serve/_private/http_proxy.py [hunk 16]
```diff
@@ -421,12 +422,20 @@ async def __call__(self, scope, receive, send):
...
+                get_random_letters(10),
+                model_id=model_id,
+                app_name=app_name,
             )
         )
+
+        start_time = time.time()
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/http_proxy.py:438]: miss moved. reverted.

**Comment by sihanwang41** [comment]
> Reference

Hi @edoakes, i removed the router & controller & deployment_state change from this pr.  Currently it is more focusing on the API change.

File: python/ray/serve/batching.py [hunk 17]
```diff
@@ -171,30 +172,6 @@ def __del__(self):
         self._handle_batch_task.cancel()
 
 
-def _extract_self_if_method_call(args: List[Any], func: Callable) -> Optional[object]:
```
Comment (sihanwang41) [inline python/ray/serve/batching.py:174]: move this function to utils.

File: python/ray/serve/api.py [hunk 18]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+                    # You can use any model loading library here
+                    # and return the loaded model. load_from_s3 is
+                    # a placeholder function.
+                    return load_from_s3(model_id)
+
+                async def __call__(self, request):
+                    # Get model id from request context
+                    model_id = serve.get_model_id()
+                    # Load model with the given model id
+                    model = await self.get_model(tag)
```
Comment (edoakes) [inline python/ray/serve/api.py:622]:
```suggestion
                    model = await self.load_model(tag)
```

File: python/ray/serve/api.py [hunk 19]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
     """
     client = get_global_client()
     client.delete_apps([name], blocking=_blocking)
+
+
+@PublicAPI(stability="alpha")
+def multiplexed(func=None, num_models_per_replica: int = 0):
+    """Coverts a function or method to a multiplexed function.
```
Comment (edoakes) [inline python/ray/serve/api.py:584]: Can you specify the considerations about `async def` vs `def` in this docstring please?

File: python/ray/serve/api.py [hunk 20]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+
+    The function can be standalone function or a method of a class. The
+    function must have at least one argument with type `str` and the argument
+    must be the first argument. The `str` argument is the model id of the model to
+    be loaded. The function can have other arguments and keyword arguments.
```
Comment (edoakes) [inline python/ray/serve/api.py:589]: It seems a little bit dangerous to me to allow other arguments and kwargs... if the resulting model is dependent on those args, then effectively our caching does not work. I would suggest we don't allow this.

File: python/ray/serve/api.py [hunk 21]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+
+    The multiplexed function will load the model with the given model id and call
+    the original function. The multiplexed function will return the return
+    value of the original function.
```
Comment (edoakes) [inline python/ray/serve/api.py:593]:
```suggestion
    The multiplexed function is called to load a model with a given model ID when
    necessary.
```

File: python/ray/serve/api.py (hunk 10 above)
Comment (edoakes) [reply python/ray/serve/api.py:586]: The return type is not `None`, it's actually `Any` -- let's make this clear @sihanwang41!

File: python/ray/serve/api.py [hunk 22]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+                    # Get model id from request context
+                    model_id = serve.get_model_id()
+                    # Load model with the given model id
+                    model = await self.get_model(tag)
+                    # Call the model
+                    return model(request)
+
+
+    Args:
+        num_models_per_replica: number of models to be loaded on each replica.
```
Comment (edoakes) [inline python/ray/serve/api.py:628]:
```suggestion
        max_models_per_replica: maximum number of models to be loaded on each replica.
```
I think we agreed this should be `max` in the design review?

File: python/ray/serve/api.py [hunk 23]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+        By default, it is 0, which means all models will be loaded on all replicas.
```
Comment (edoakes) [inline python/ray/serve/api.py:629]:
```suggestion
        By default, it is 0, which means there is no limit.
```

Comment (edoakes) [reply python/ray/serve/api.py:629]: nit: this should be `-1` (standard convention)

File: python/ray/serve/api.py [hunk 24]
```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
...
+
+    return _multiplex_decorator(func) if callable(func) else _multiplex_decorator
+
+
+@PublicAPI(stability="alpha")
+def get_model_id() -> str:
+    """Returns the model id of the current request.
+
+    When user defines a multiplexed deployment, the model id of the current request
+    can be retrieved by calling `serve.get_model_id()`.
```
Comment (edoakes) [inline python/ray/serve/api.py:684]: add usage example to docstring please!

**Review by edoakes — COMMENTED** [review]
Only reviewed the API so far. These will need to be added to the API reference as well, should go under some "experimental" heading IMO. That can be done once the full functionality is merged and a user guide is added.

File: python/ray/serve/api.py (hunk 18 above)
Comment (edoakes) [reply python/ray/serve/api.py:622]: (doesn't match fn name right now)

**Comment by sihanwang41** [comment]
ping ^^ @edoakes

File: python/ray/serve/api.py (hunk 11 above)
Comment (edoakes) [reply python/ray/serve/api.py:698]: @sihanwang41 can we call this `get_multiplexed_model_id()` to make it clear it's used for multiplexing?

File: python/ray/serve/handle.py [hunk 25]
```diff
@@ -339,6 +314,7 @@ def options(
         self,
         *,
         method_name: Union[str, DEFAULT] = DEFAULT.VALUE,
+        model_id: Union[str, DEFAULT] = DEFAULT.VALUE,
```
Comment (edoakes) [inline python/ray/serve/handle.py:317]: I thought handle changes are reverted from this pr?

File: python/ray/serve/api.py [hunk 26]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+@PublicAPI(stability="alpha")
+def multiplexed(
+    func: Optional[Callable[..., Any]] = None, max_num_models_per_replica: int = -1
+):
+    """[EXPERIMENTAL] Coverts a function or method to a multiplexed function.
+
+    The function can be standalone function or a method of a class. The
+    function must have exact one argument with type `str` and the argument
+    must be the first argument. The `str` argument is the model id of the model to
+    be loaded.
```
Comment (zcin) [inline python/ray/serve/api.py:591]: This is a bit confusing to me. Why mention it must be the first argument if it can only take one argument? Lmk if I'm interpreting this incorrectly.
```suggestion
    function must have exactly one argument, the model id of type `str` for the
    model to be loaded.
```

File: python/ray/serve/api.py [hunk 27]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+    a `__del__` method in your model class. The `__del__` method will be called when
+    the model is unloaded.
+
+    Example:
+
+    .. code-block:: python
+            from ray import serve
+
+            @serve.deployment
+            class MultiplexedDeployment:
```
Comment (edoakes) [inline python/ray/serve/api.py:613]: suggest to add constructor that sets some "s3 path" or something (to show that you can still do that)

File: python/ray/serve/api.py [hunk 28]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+    """[EXPERIMENTAL] Coverts a function or method to a multiplexed function.
+
+    The function can be standalone function or a method of a class. The
+    function must have exact one argument with type `str` and the argument
+    must be the first argument. The `str` argument is the model id of the model to
+    be loaded.
+
+    User have to define the function with `async def` and the function must be
+    an async function. It is recommended to define courotine for long running
+    IO tasks in the function, it can prevent blocking the event loop.
```
Comment (zcin) [inline python/ray/serve/api.py:595]:
```suggestion
    an async function. It is recommended to define coroutines for long running
    IO tasks in the function to avoid blocking the event loop.
```

File: python/ray/serve/api.py [hunk 29]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+
+    The multiplexed function is called to load a model with a given model ID when
+    necessary.
+
+    When the number of models in one replica is larger than max_num_models_per_replica,
+    the multiplexed function will unload the model with LRU policy.
```
Comment (edoakes) [inline python/ray/serve/api.py:601]:
```suggestion
    When the number of models in one replica is larger than max_num_models_per_replica,
    the models will be unloaded using an LRU policy.
```

**Review by edoakes — COMMENTED** [review]
@sihanwang41 I thought the intention was to only include the API changes in this first PR?

File: python/ray/serve/api.py [hunk 30]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+    The function can be standalone function or a method of a class. The
+    function must have exact one argument with type `str` and the argument
+    must be the first argument. The `str` argument is the model id of the model to
+    be loaded.
+
+    User have to define the function with `async def` and the function must be
+    an async function. It is recommended to define courotine for long running
+    IO tasks in the function, it can prevent blocking the event loop.
+
+    The multiplexed function is called to load a model with a given model ID when
```
Comment (zcin) [inline python/ray/serve/api.py:597]:
```suggestion
    The multiplexed function is called to load a model with the given model ID when
```

File: python/ray/serve/api.py [hunk 31]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+    necessary.
+
+    When the number of models in one replica is larger than max_num_models_per_replica,
+    the multiplexed function will unload the model with LRU policy.
+
+    If you want to release resource after the model is loaded, you can define
```
Comment (zcin) [inline python/ray/serve/api.py:603]:
```suggestion
    If you want to release resources after the model is loaded, you can define
```

File: python/ray/serve/api.py [hunk 32]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+        By default, it is -1, which means there is no number limit to load models.
+    """
+    if func is not None:
+        if not callable(func):
+            raise TypeError(
+                "@serve.multiplex can only be used to decorate functions or methods."
+            )
+
+    if max_num_models_per_replica != -1 and max_num_models_per_replica <= 0:
+        raise ValueError("max_num_models_per_replica must be larger than 0.")
```
Comment (zcin) [inline python/ray/serve/api.py:643]:
```suggestion
        raise ValueError("max_num_models_per_replica must be positive.")
```

File: python/ray/serve/api.py [hunk 33]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+            )
+
+        @wraps(func)
+        async def _multiplex_wrapper(*args):
+            self = _extract_self_if_method_call(args, func)
+            signature = inspect.signature(func)
+            if self is None:
+                if len(signature.parameters) != 1:
+                    raise ValueError(
+                        "The function must have exact one argument with type `str` "
```
Comment (zcin) [inline python/ray/serve/api.py:668]:
```suggestion
                        "The function must have exactly one argument with type `str` "
```

File: python/ray/serve/api.py [hunk 34]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+                        "as model_id."
+                    )
+                multiplex_object = func
+                model_id = args[0]
+            else:
+                # count self as an argument
+                if len(signature.parameters) != 2:
+                    raise ValueError(
+                        "The method must have exact one argument with type `str` "
```
Comment (zcin) [inline python/ray/serve/api.py:677]:
```suggestion
                        "The method must have exactly one argument with type `str` "
```

File: python/ray/serve/multiplex.py [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+
+        The model multiplexer is a wrapper class that wraps the model load function
+        and provides the LRU caching functionality, and the model load function should
+        be a coroutine function that takes the model ID as the first argument and
+        returns the model handle.
+        The model multiplexer will also ensure that the number of models on the current
+        replica does not exceed the specified limit.
+        The model will be unloaded in the LRU order, the model multiplexer will call the
+        model's __del__ attribute if it exists (e.g., PyTorch models) to clean up the
+        model resources eargerly.
```
Comment (zcin) [inline python/ray/serve/multiplex.py:32]:
```suggestion
        model resources eagerly.
```

File: python/ray/serve/api.py [hunk 36]
```diff
@@ -571,3 +577,153 @@ def delete(name: str, _blocking: bool = True):
...
+                    # Get model id from request context
+                    model_id = serve.get_model_id()
+                    # Load model with the given model id
+                    model = await self.load_model(model_id)
+                    # Call the model
+                    return model(request)
+
+
+    Args:
+        max_num_models_per_replica: number of models to be loaded on each replica.
```
Comment (zcin) [inline python/ray/serve/api.py:633]:
```suggestion
        max_num_models_per_replica: the maximum number of models to be loaded on each replica.
```

**Review by zcin — COMMENTED** [review]
Some wording/docstring nits

**Comment by sihanwang41** [comment]
> @sihanwang41 I thought the intention was to only include the API changes in this first PR?

trimmed!

File: python/ray/serve/api.py [hunk 37]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
     """
     client = get_global_client()
     client.delete_apps([name], blocking=_blocking)
+
+
+@PublicAPI(stability="alpha")
+def multiplexed(
+    func: Optional[Callable[..., Any]] = None, max_num_models_per_replica: int = -1
+):
+    """[EXPERIMENTAL] Converts a function or method to a multiplexed function.
```
Comment (edoakes) [inline python/ray/serve/api.py:581]:
```suggestion
    """[EXPERIMENTAL] Defines a function or method used to load multiplexed models in a replica.
```

**Review by edoakes — COMMENTED** [review]
In the future please do a self review before pinging so I don't need to catch things like mis-typed API names in docstrings

File: python/ray/serve/api.py (hunk 27 above)
Comment (sihanwang41) [reply python/ray/serve/api.py:608]: Hi @edoakes added a constructor for it in the constructor.

File: python/ray/serve/api.py [hunk 38]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+                    # a placeholder function.
+                    return load_from_s3(model_id)
+
+                async def __call__(self, request):
+                    # Get model id from request context
+                    model_id = serve.get_multiplexed_model_id()
+                    # Load model with the given model id
+                    model = await self.load_model(model_id)
+                    # Call the model
+                    return model(request)
```
Comment (edoakes) [inline python/ray/serve/api.py:628]:
```suggestion
                    return model(request)
```
obvious, too many comments makes it harder to read IMO

File: python/ray/serve/api.py [hunk 39]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+            @serve.deployment
+            class MultiplexedDeployment:
+
+                def __init__(self):
+                    # Define s3 base path to load models
+                    self.s3_base_path = "s3://my_bucket/my_models"
+
+                @serve.multiplexed(max_num_models_per_replica=5)
+                async def load_model(self, model_id: str) -> Any:
+                    # Load model with the given tag
```
Comment (edoakes) [inline python/ray/serve/api.py:616]:
```suggestion
```
also obvious from the code

File: python/ray/serve/api.py [hunk 40]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+    Example:
+
+    .. code-block:: python
+            from ray import serve
+
+            @serve.deployment
+            class MultiplexedDeployment:
+
+                def __init__(self):
+                    # Define s3 base path to load models
```
Comment (edoakes) [inline python/ray/serve/api.py:611]:
```suggestion
                    # Define s3 base path to load models.
```
User-facing, use proper grammar/punctuation :)

File: python/ray/serve/api.py [hunk 41]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+                @serve.multiplexed(max_num_models_per_replica=5)
+                async def load_model(self, model_id: str) -> Any:
+                    # Load model with the given tag
+                    # You can use any model loading library here
+                    # and return the loaded model. load_from_s3 is
+                    # a placeholder function.
+                    return load_from_s3(model_id)
+
+                async def __call__(self, request):
+                    # Get model id from request context
```
Comment (edoakes) [inline python/ray/serve/api.py:623]:
```suggestion
                    # Get the model_id from the request context.
```

File: python/ray/serve/api.py [hunk 42]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+                    model_id = serve.get_multiplexed_model_id()
+                    # Load model with the given model id
```
Comment (edoakes) [inline python/ray/serve/api.py:625]:
```suggestion
                    # Load the model for the requested model_id.
                    # If the model is already cached locally, this will just be a dictionary lookup.
```

File: python/ray/serve/api.py [hunk 43]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+
+    raise NotImplementedError("Multiplexed deployment is not supported yet.")
+
+
+@PublicAPI(stability="alpha")
+def get_multiplexed_model_id() -> str:
+    """[EXPERIMENTAL] Returns the model id of the current request.
+
+    When user defines a multiplexed deployment, the model id of the current request
+    can be retrieved by calling `serve.get_multiplexed_model_id()`.
```
Comment (edoakes) [inline python/ray/serve/api.py:645]:
```suggestion
    """[EXPERIMENTAL] Get the multiplexed model ID for the current request.

    This is used with a function decorated with `@serve.multiplexed` to cache multiple independent
    models within a single deployment.
```

File: python/ray/serve/api.py [hunk 44]
```diff
@@ -571,3 +572,100 @@ def delete(name: str, _blocking: bool = True):
...
+
+    .. code-block:: python
+            from ray import serve
+
+            @serve.deployment
+            class MultiplexedDeployment:
```
Comment (edoakes) [inline python/ray/serve/api.py:651]: probably don't need to duplicate the entire class example here, maybe just show the `__call__` part

we should also describe where the ID comes from... I'd say this sample should show the client side setting a model ID and the replica getting it (in comment say "this will return 'model-abc' that we set in the client"

**Review by edoakes — COMMENTED** [review]
Mostly nits + would like to improve the example for `get_multiplexed_model_id`

File: python/ray/serve/api.py [hunk 45]
```diff
@@ -571,3 +572,97 @@ def delete(name: str, _blocking: bool = True):
...
+            from ray import serve
+            import ray
+            import requests
+            # client code
+            # User needs to set the model id in the request header with the key
+            # "ray_serve_request_model_id" when sending requests to the http proxy.
+            requests.get("http://localhost:8000",
+                headers={"ray_serve_request_model_id": "model_1"})
+            # Or user can also set model id inside the serve handle.
+            handle.options(model_id="model_1").remote("blablabla")
```
Comment (edoakes) [inline python/ray/serve/api.py:661]:
```suggestion
            requests.get("http://localhost:8000",
                headers={"ray_serve_multiplexed_model_id": "model_1"})
            # Or user can also set model id inside the serve handle.
            handle.options(multiplexed_model_id="model_1").remote("blablabla")
```
let's please keep the naming consistent across the API

Comment (edoakes) [reply python/ray/serve/api.py:661]: we might not need the `ray_serve_` prefix in the header, i don't have too much preference

File: python/ray/serve/api.py [hunk 46]
```diff
@@ -571,3 +572,97 @@ def delete(name: str, _blocking: bool = True):
...
+            # User needs to set the model id in the request header with the key
+            # "ray_serve_multiplexed_model_id" when sending requests to the http proxy.
+            requests.get("http://localhost:8000",
+                headers={"ray_serve_multiplexed_model_id": "model_1"})
+            # Or user can also set model id inside the serve handle.
+            handle.options(multiplexed_model_id="model_1").remote("blablabla")
+
+            # In replica, You can retrieve the model id from the
+            # request context.
+            assert serve.get_multiplexed_model_id() == "model_1"
```
Comment (edoakes) [inline python/ray/serve/api.py:665]:
```suggestion
    .. code-block:: python
            import ray
            from ray import serve
            import requests
            
            # Set the multiplexed model id in the request header in the client:
            # "ray_serve_multiplexed_model_id" when sending requests to the http proxy.
            requests.get("http://localhost:8000",
                headers={"ray_serve_multiplexed_model_id": "model_1"})
                
            # This can also be set when using a `RayServeHandle`.
            handle.options(multiplexed_model_id="model_1").remote()

            # Then when inside your deployment code, you can retrieve the ID using
            # `get_multiplexed_model_id()`:
            def my_deployment_function(request):
                assert serve.get_multiplexed_model_id() == "model_1"
```

- This is user-facing, please make sure it's highly polished (punctuation, consistent usage of concept names).
- Use active voice and present tense, don't reference "the user" in documentation.

File: python/ray/serve/api.py (hunk 45 above)
Comment (sihanwang41) [reply python/ray/serve/api.py]: I think it is more readable to the user for setting up the headers. i will keep it as it is right now.

File: python/ray/serve/api.py (hunk 46 above)
Comment (sihanwang41) [reply python/ray/serve/api.py]: Read it another round, update the doc examples, looks better now :)

File: python/ray/serve/api.py [hunk 47]
```diff
@@ -571,3 +572,98 @@ def delete(name: str, _blocking: bool = True):
...
+            # "ray_serve_multiplexed_model_id" in the request
+            # headers when sending requests to the http proxy.
+            requests.get("http://localhost:8000",
+                headers={"ray_serve_multiplexed_model_id": "model_1"})
+            # This can also be set when using `RayServeHandle`.
+            handle.options(multiplexed_model_id="model_1").remote("blablabla")
+
+            # In your deployment code, you can retrieve the model id from
+            # `get_multiplexed_model_id()`.
+            def my_deployment_function(request):
```
Comment (edoakes) [inline python/ray/serve/api.py:666]:
```suggestion
            # In your deployment code, you can retrieve the model id from
            # `get_multiplexed_model_id()`.
            @serve.deployment
            def my_deployment_function(request):
```
one more nit to make it even clearer :)

**Review by edoakes — APPROVED** [review]
